keps/sig-auth/0000-20170814-bounding-self-labeling-kubelets.md

```diff
@@ -0,0 +1,163 @@
+trusted server certificates. They use the secret mechanism to distribute the
+serving certificate key. An intruder captures the dedicated nginx workload in
+the same way and can now use the node certificate to read the company's serving
+certificate key.
+
+## Proposal
+
+The NodeRestriction admission plugin will be updated to limit the labels a kubelet
+can set on its Node object. Kubelets will only be allowed to set the following labels.
+
+### Well-known label keys
+
+This list includes all of the labels that kubelets already set on their Node objects.
+
+```
+kubernetes.io/hostname
+failure-domain.beta.kubernetes.io/zone
+failure-domain.beta.kubernetes.io/region
+beta.kubernetes.io/instance-type
+beta.kubernetes.io/os
+beta.kubernetes.io/arch
+```
+
+### Well-known kubelet-related label namespace
+
+This prefix allows kubelets to control arbitrary labels under their own label namespace.
+
+```
+[*.]kubelet.kubernetes.io/*
+```
+
+### Well-known topology-related label namespace
+
+This prefix allows kubelets to control arbitrary labels related to topology.
+
+```
+[*.]topology.kubernetes.io/*
+```
+
+#### CSI use case
+
+CSI allows drivers to provide [arbitrary topology labels](https://github.com/container-storage-interface/spec/blob/v0.3.0/csi.proto#L524-L549),
+which are planned to be added as labels to the Node by the kubelet.
+This section describes how CSI could continue to support arbitrary topology information provided by drivers,
+and conform to a consistent topology-related prefix when adding labels to Nodes.
+
+The CSI topology labels are in the form `<name>` or `<prefix>/<name>`, where both prefix and name are up to 63 characters long.
+
+Kubernetes labels allow a prefix up to [253 characters long](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).
+
+CSI topology labels could be transformed to kubernetes labels within the permitted topology label namespace:
+
+* CSI topology labels that did not have a prefix would have a prefix of `topology.kubernetes.io/` prepended
+* CSI topology labels with a prefix not ending in `topology.kubernetes.io` would have that string added to their prefix
+
+Component behavior:
+
+* The kubelet would transform CSI topology to kubernetes labels before adding them to the `Node` API object.
+* The CSI provisioner would transform CSI topology labels to kubernetes labels when [retrieving topology info from a given node](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L295-L305).
+* The CSI provisioner would transform kubernetes labels into CSI topology labels when [building required topology from a storage class AllowedTopologies](https://github.com/kubernetes-csi/external-provisioner/blob/v0.5.0-alpha.0/pkg/controller/topology.go#L75-L91).
+* The associated storage class would continue to specify topology in terms of kubernetes label selectors.
+
```

Comment:
nit: for the sake of completeness, the provisioner also translates CSI topology from the provisioned volume to volume node affinity, which uses k8s labels.